python/src/iceberg/schema.py

```diff
@@ -0,0 +1,448 @@
+# under the License.
+
+from __future__ import annotations
+
+import sys
+from abc import ABC, abstractmethod
+from typing import Dict, Generic, Iterable, List, TypeVar
+
+if sys.version_info >= (3, 8):
+    from functools import singledispatch  # pragma: no cover
+else:
+    from singledispatch import singledispatch  # pragma: no cover
+
+from iceberg.types import (
+    IcebergType,
+    ListType,
+    MapType,
+    NestedField,
+    PrimitiveType,
+    StructType,
+)
+
+T = TypeVar("T")
+
+
+class Schema:
+    """A table Schema
+
+    Example:
+        >>> from iceberg import schema
+        >>> from iceberg import types
+    """
+
+    def __init__(self, *columns: Iterable[NestedField], schema_id: int, identifier_field_ids: List[int] = []):
+        self._struct = StructType(*columns)  # type: ignore
+        self._schema_id = schema_id
+        self._identifier_field_ids = identifier_field_ids
+        self._name_to_id: Dict[str, int] = index_by_name(self)
+        self._name_to_id_lower: Dict[str, int] = {}  # Should be accessed through self._lazy_name_to_id_lower()
+        self._id_to_field: Dict[int, NestedField] = {}  # Should be accessed through self._lazy_id_to_field()
+
+    def __str__(self):
+        return "table {\n" + "\n".join(["  " + str(field) for field in self.columns]) + "\n}"
+
+    def __repr__(self):
+        return (
+            f"Schema(fields={repr(self.columns)}, schema_id={self.schema_id}, identifier_field_ids={self.identifier_field_ids})"
+        )
+
+    @property
+    def columns(self) -> Iterable[NestedField]:
+        """A list of the top-level fields in the underlying struct"""
+        return self._struct.fields
+
+    @property
+    def schema_id(self) -> int:
+        """The ID of this Schema"""
+        return self._schema_id
+
+    @property
+    def identifier_field_ids(self) -> List[int]:
+        return self._identifier_field_ids
+
+    def _lazy_id_to_field(self) -> Dict[int, NestedField]:
+        """Returns an index of field ID to NestedField instance
+
+        This property is calculated once when called for the first time. Subsequent calls to this property will use a cached index.
+        """
+        if not self._id_to_field:
+            self._id_to_field = index_by_id(self)
+        return self._id_to_field
+
+    def _lazy_name_to_id_lower(self) -> Dict[str, int]:
+        """Returns an index of lower-case field names to field IDs
+
+        This property is calculated once when called for the first time. Subsequent calls to this property will use a cached index.
+        """
+        if not self._name_to_id_lower:
+            self._name_to_id_lower = {name.lower(): field_id for name, field_id in self._name_to_id.items()}
+        return self._name_to_id_lower
+
+    def as_struct(self) -> StructType:
+        """Returns the underlying struct"""
+        return self._struct
+
+    def find_field(self, name_or_id: str | int, case_sensitive: bool = True) -> NestedField:
```

Comment:
I know I suggested this syntax, but I had IntelliJ complain that python 3.7 doesn't support it when I tried to use it. But tests are passing so it's okay?

Reply:
It must not be recognizing the `from __future__ import annotations` at the top which is required to backport it to 3.7, but yeah if the tests are passing we're ok.

Reply:
I found this when I tried to use it in my recent changes and didn't know to import from __future__. I'll update that eventually.